xayn-ai/src/ranker/public.rs

```diff
@@ -0,0 +1,166 @@
+            user_interests: UserInterests::default(),
+        }
+    }
+}
+
+impl<SV, SM, KV, KM> Builder<SV, SM, KV, KM> {
+    /// Sets the serialized state to use.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the state cannot be deserialized.
+    pub fn with_serialized_state(mut self, bytes: impl AsRef<[u8]>) -> Result<Self, Error> {
+        self.user_interests = bincode::deserialize(bytes.as_ref())?;
+        Ok(self)
+    }
+
+    /// Sets the reader of the vocabulary, and bert model used to initialize the SMBert.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the data cannot be deserialized.
+    pub fn with_smbert_from_reader<W, N>(self, vocab: W, bert: N) -> Builder<W, N, KV, KM> {
+        Builder {
+            smbert: SMBertBuilder::new(vocab, bert),
+            kpe: self.kpe,
+            user_interests: self.user_interests,
+        }
+    }
+
+    /// Sets the path of the vocabulary, and bert model used to initialize the SMBert.
+    ///
+    /// # Errors
+    ///
+    /// Fails if a file does not exist or if the data cannot be deserialized.
+    pub fn with_smbert_from_file(
+        self,
+        vocab: impl AsRef<Path>,
+        bert: impl AsRef<Path>,
+    ) -> Result<Builder<impl BufRead, impl Read, KV, KM>, Error> {
+        Ok(Builder {
+            smbert: SMBertBuilder::from_files(vocab, bert)?,
+            kpe: self.kpe,
+            user_interests: self.user_interests,
+        })
+    }
+
+    /// Sets the reader of the vocabulary, bert model, CNN and classifier used
+    /// to initialize the KPE.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the data cannot be deserialized.
+    pub fn with_kpe_from_reader<W, N>(
+        self,
+        vocab: W,
+        bert: N,
+        cnn: impl Read,
+        classifier: impl Read,
+    ) -> Result<Builder<SV, SM, W, N>, Error> {
+        Ok(Builder {
+            smbert: self.smbert,
+            kpe: KPEBuilder::new(
+                vocab,
+                bert,
+                BinParams::deserialize_from(cnn)?,
+                BinParams::deserialize_from(classifier)?,
+            ),
+            user_interests: self.user_interests,
+        })
+    }
+
+    /// Sets the path of the vocabulary, bert model, CNN and classifier used
+    /// to initialize the KPE.
+    ///
+    /// # Errors
+    ///
+    /// Fails if a file does not exist or if the data cannot be deserialized.
+    pub fn with_kpe_from_file(
+        self,
+        vocab: impl AsRef<Path>,
+        bert: impl AsRef<Path>,
+        cnn: impl AsRef<Path>,
+        classifier: impl AsRef<Path>,
+    ) -> Result<Builder<SV, SM, impl BufRead, impl Read>, Error> {
+        Ok(Builder {
+            smbert: self.smbert,
+            kpe: KPEBuilder::from_files(vocab, bert, cnn, classifier)?,
+            user_interests: self.user_interests,
+        })
+    }
+
+    /// Creates a `Reranker`.
+    ///
+    /// # Errors
+    ///
+    /// Fails if the SMBert or KPE cannot be initialized.
```

Comment:
```suggestion
    /// Fails if the SMBert or KPE cannot be initialized.
    ///
    /// For example because reading from a file failed or the read bytes are in an unexpected format.
```

Reply:
maybe, or something similar.